File: pythonCode/kelvinFilter.py
import cv2
import numpy as np
import sys

#read input image
image = cv2.imread("../assets/anish.jpg")

#check if image exists
if image is None:
    print("can not find image")
    sys.exit()

#create a copy of input image to work on
output = image.copy()

#split the channels
blueChannel, greenChannel, redChannel = cv2.split(output)

#Interpolation values
redValuesOriginal = np.array([0, 60, 110, 150, 235, 255])
redValues =         np.array([0, 102, 185, 220, 245, 245 ])
greenValuesOriginal = np.array([0, 68, 105, 190, 255])
greenValues =         np.array([0, 68, 120, 220, 255 ])
blueValuesOriginal = np.array([0, 88, 145, 185, 255])
blueValues =         np.array([0, 12, 140, 212, 255])

#create lookup table
allValues = np.arange(0, 256)
#Creating the lookuptable for blue channel
blueLookuptable = np.interp(allValues, blueValuesOriginal, blueValues)
#Creating the lookuptable for green channel
greenLookuptable = np.interp(allValues, greenValuesOriginal, greenValues)
#Creating the lookuptable for red channel
redLookuptable = np.interp(allValues, redValuesOriginal, redValues)

#Apply the mapping for blue channel
blueChannel = cv2.LUT(blueChannel, blueLookuptable)
#Apply the mapping for green channel
greenChannel = cv2.LUT(greenChannel, greenLookuptable)
#Apply the mapping for red channel
redChannel = cv2.LUT(redChannel, redLookuptable)

#merging back the channels
output = cv2.merge([blueChannel, greenChannel, redChannel])

#convert to uint8
output = np.uint8(output)

#create windows to display images
cv2.namedWindow("image", cv2.WINDOW_AUTOSIZE)
cv2.namedWindow("Kelvin", cv2.WINDOW_AUTOSIZE)

#display images
cv2.imshow("image", image)
cv2.imshow("Kelvin", output)

#press esc to exit the program
cv2.waitKey(0)

#destroy all windows
cv2.destroyAllWindows()

File: cPlusPlusCode/kelvinFilter.cpp
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<iostream>

using namespace cv;
using namespace std;

//interpolation
void interpolation(uchar* lut, float* curve, float* originalValue){
    for(int i=0; i<256; i++){
        int j=0;
        float a = i;
        while (a>originalValue[j]){
            j++;
        }
        if(a == originalValue[j]){
            lut[i] = curve[j];
            continue;
        }
        float slope = ((float)(curve[j] - curve[j-1]))/((float)(originalValue[j] - originalValue[j-1]));
        float constant = curve[j] - slope * originalValue[j];
        lut[i] = slope * a + constant;
    }
}

int main(){

    //Read input image
    Mat image = imread("../assets/anish.jpg");

    //Check is image exits
    if(image.empty()){
        cout<<"can not find image"<<endl;
        return 0;
    }

    //define variable to store output image
    Mat output;

    //create a clone of input image to work on
    output = image.clone();

    //Split the channels
    vector<Mat> channels;
    split(output, channels);

    // Interpolation values
    float redValuesOriginal[] = {0, 60, 110, 150, 235, 255};
    float redValues[] =         {0, 102, 185, 220, 245, 245};
    float greenValuesOriginal[] = {0, 68, 105, 190, 255};
    float greenValues[] =         {0, 68, 120, 220, 255};
    float blueValuesOriginal[] = {0, 88, 145, 185, 255};
    float blueValues[] =         {0, 12, 140, 212, 255};

    //create lookup table
    Mat lookupTable(1, 256, CV_8U);
    uchar* lut = lookupTable.ptr();

    //Apply interpolation and create lookup table
    interpolation(lut, blueValues, blueValuesOriginal);
    //Apply mapping for Blue Channel
    LUT(channels[0], lookupTable, channels[0]);

    //Apply interpolation and create lookup table
    interpolation(lut, greenValues, greenValuesOriginal);
    //Apply mapping for green Channel
    LUT(channels[1], lookupTable, channels[1]);

    //Apply interpolation and create lookup table
    interpolation(lut, redValues, redValuesOriginal);
    //Apply mapping for Red Channel
    LUT(channels[2], lookupTable, channels[2]);

    //Merge back the channels
    merge(channels, output);

    //create windows to display images
    namedWindow("image", WINDOW_AUTOSIZE);
    namedWindow("Kelvin", WINDOW_AUTOSIZE);

    //Display images
    imshow("image", image);
    imshow("Kelvin", output);

    //Press esc to exi the program
    waitKey(0);

    //Close all the opened windows
    destroyAllWindows();

    return 0;
}